### PR TITLE
pkg/fileutil: lock file on Windows

### PR DESCRIPTION
**Pull request by gyuho**
Fix the following:

1. windows: fix WAL lock https://github.com/coreos/etcd/issues/1972
2. etcdserver: failed to purge snap file https://github.com/coreos/etcd/issues/2515

Tested in Windows machine and tests passed.

/cc @heyitsanthony @xiang90 

Thanks.

### CODE REVIEW

File: pkg/fileutil/readdir_windows.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+package fileutil
+
+import (
+	"os"
+	"path/filepath"
+	"sort"
+)
+
+func ReadDir(dir string) ([]string, error) {
```
Comment (heyitsanthony) [inline pkg/fileutil/readdir_windows.go]: Is windows generating extra files? NFS has an issue with `.nfs` lock files...

Both windows and linux should probably do the same thing here-- get the directory file names from `os.Readdirnames(-1)` and filter out the dot files.

Comment (gyuho) [inline pkg/fileutil/readdir_windows.go]: Yeah it was reading the locked file even after it deletes that file.  I will double check. Thanks

File: pkg/fileutil/lock_windows.go [hunk 2]
```diff
@@ -16,17 +16,112 @@
...
+
+func lockFile(fd syscall.Handle, flags uint32) error {
+	var flag uint32 = LOCKFILE_EXCLUSIVE_LOCK
+	if flags != 0 {
+		flag |= flags
+	}
+	if fd == syscall.InvalidHandle {
+		return nil
+	}
+	for {
```
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: why does this need a retry loop?

File: pkg/fileutil/fileutil_test.go [hunk 3]
```diff
@@ -41,10 +42,10 @@ func TestIsDirWriteable(t *testing.T) {
 		// http://stackoverflow.com/questions/20609415/cross-compiling-user-current-not-implemented-on-linux-amd64
 		t.Skipf("failed to get current user: %v", err)
 	}
-	if me.Name == "root" || me.Name == "Administrator" {
+	if me.Name == "root" || me.Name == "Administrator" || runtime.GOOS == "windows" {
```
Comment (heyitsanthony) [inline pkg/fileutil/fileutil_test.go]: `me.Name == "root" || runtime.GOOS == "windows" `?

**Comment by gyuho** [comment]
@heyitsanthony PTAL.

1. removed `time.Sleep` (not needed). I tested without it and worked.
2. added comment about why we need separate `ReadDir` function for Windows, which returns outdated files in some systems (it's `nil`)
3. removed `Administrator` checking in `fileutil_test.go`. Windows doesn't support Chmod, so we just skip it.

Thanks!

File: pkg/fileutil/readdir_windows.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	visit := func(path string, f os.FileInfo, err error) error {
+		if f != nil {
+			fp2 := filepath.Base(filepath.Dir(path))
+			if fp1 == fp2 { // to skip walking sub-directories like Readdir
+				fnames = append(fnames, f.Name())
+			}
+		}
+		return nil
+	}
+	if err := filepath.Walk(dir, visit); err != nil {
```
Comment (heyitsanthony) [inline pkg/fileutil/readdir_windows.go]: Would `Readdir` work instead? `Walk` seems somewhat extreme.

Comment (gyuho) [inline pkg/fileutil/readdir_windows.go]: I tried `Readdir`, basically same as `Readdirnames` (which uses `Readdir`) and it returns `nil` files in Windows as well.

Comment (heyitsanthony) [inline pkg/fileutil/readdir_windows.go]: OK, I looked at the go sources to figure out what's happening here. It seems the main difference between `readdirnames`/`readdir` and `walk` is `walk` is `lstat`ing the files after it gets the names so there could be some race happening and this is masking it.

Is this so the PurgeFile tests pass or is something else failing too?

Comment (gyuho) [inline pkg/fileutil/readdir_windows.go]: Only this fails https://github.com/coreos/etcd/blob/master/pkg/fileutil/purge_test.go#L146. `5.test` is `nil` (I checked manually as well), but it is returned in `Readdir`.

File: pkg/fileutil/lock_windows.go [hunk 5]
```diff
@@ -16,17 +16,106 @@
...
 
 func LockFile(path string, flag int, perm os.FileMode) (*LockedFile, error) {
-	// TODO make this actually work
-	f, err := os.OpenFile(path, flag, perm)
+	f, err := open(path, flag, perm)
 	if err != nil {
 		return nil, err
 	}
+	if err := lockFile(syscall.Handle(f.Fd()), 0); err != nil {
+		return nil, err
```
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: `f.Close()`?

File: pkg/fileutil/lock_windows.go [hunk 6]
```diff
@@ -16,17 +16,106 @@
...
+)
 
 func TryLockFile(path string, flag int, perm os.FileMode) (*LockedFile, error) {
-	return LockFile(path, flag, perm)
+	f, err := open(path, flag, perm)
+	if err != nil {
+		return nil, err
+	}
+	if err := lockFile(syscall.Handle(f.Fd()), LOCKFILE_FAIL_IMMEDIATELY); err != nil {
+		return nil, err
```
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: `f.Close()`?

File: pkg/fileutil/lock_windows.go [hunk 7]
```diff
@@ -16,17 +16,108 @@
...
+)
 
 func TryLockFile(path string, flag int, perm os.FileMode) (*LockedFile, error) {
-	return LockFile(path, flag, perm)
+	f, err := open(path, flag, perm)
+	if err != nil {
+		return nil, err
+	}
+	if err := lockFile(syscall.Handle(f.Fd()), LOCKFILE_FAIL_IMMEDIATELY); err != nil {
+		f.Close()
```
Comment (gyuho) [inline pkg/fileutil/lock_windows.go]: @heyitsanthony  Just added `Close`s. Thanks.

File: pkg/fileutil/lock_windows.go [hunk 8]
```diff
@@ -16,17 +16,106 @@
...
+		0)
+	if err != nil {
+		return nil, err
+	}
+	return os.NewFile(uintptr(fd), path), nil
+}
+
+func lockFile(fd syscall.Handle, flags uint32) error {
+	var flag uint32 = LOCKFILE_EXCLUSIVE_LOCK
+	if flags != 0 {
```
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: unnecessary `if`?

Comment (gyuho) [inline pkg/fileutil/lock_windows.go]: you are right. Just removed. Thanks!

File: pkg/fileutil/lock_windows.go (hunk 7 above)
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: This didn't happen to fix the readdir thing, did it? Windows might be picky about fd's hanging around.

**Comment by gyuho** [comment]
@heyitsanthony wow you are totally right. After adding `f.Close`, tests pass. Just fixed back to use `ReadDir`. Thanks!

File: pkg/fileutil/lock_windows.go [hunk 9]
```diff
@@ -16,17 +16,106 @@
...
+	}
 	return &LockedFile{f}, nil
 }
+
+func open(path string, flag int, perm os.FileMode) (*os.File, error) {
+	if path == "" {
+		return nil, fmt.Errorf("cannot open empty filename")
+	}
+	var access uint32
+	switch flag & (syscall.O_RDONLY | syscall.O_WRONLY | syscall.O_RDWR) {
```
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: `panic` on  `default:` instead of masking out flags?

**Comment by heyitsanthony** [comment]
OK lgtm aside from the open flag stuff. Thanks!

File: pkg/fileutil/lock_windows.go [hunk 10]
```diff
@@ -16,17 +16,108 @@
...
+	case syscall.O_RDONLY:
+		access = syscall.GENERIC_READ
+	case syscall.O_WRONLY:
+		access = syscall.GENERIC_WRITE
+	case syscall.O_RDWR:
+		access = syscall.GENERIC_READ | syscall.GENERIC_WRITE
+	default:
+		panic(fmt.Errorf("flag %v is not supported", flag))
```
Comment (gyuho) [inline pkg/fileutil/lock_windows.go]: just added. Will merge after green lights. Thanks!!

File: pkg/fileutil/lock_windows.go (hunk 9 above)
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: This should be `switch flag {`...?

File: pkg/fileutil/lock_windows.go (hunk 10 above)
Comment (gyuho) [inline pkg/fileutil/lock_windows.go]: @heyitsanthony just fixed to `switch flag`. Thanks for double-checking!

Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: @gyuho you may want to confirm `etcd` runs as expected on windows. There's an `O_CREATE` flag that may cause this to panic (`flag & O_CREATE`, I guess?)

Comment (gyuho) [inline pkg/fileutil/lock_windows.go]: Yeah actually just tested and found the panic. So I added another check with `syscall.O_WRONLY | syscall.O_CREAT`. Now works.

Thanks!

**Pull request merged**